### PR TITLE
MAINT: Change the type of error in numpy.ndarray.fill

### PR DESCRIPTION
**Pull request by kianasun**
I think it is a ``TypeError`` when the value is not a scalar.

### CODE REVIEW

**Comment by eric-wieser** [comment]
A change like this needs a test, if only so we can see when the behavior changes

**Comment by kianasun** [comment]
Hi @eric-wieser , I added tests for it. I also updated the document because it seems that `value` does not need to be a scalar value when the `dtype` of `a` is `"O"`.

File: numpy/core/_add_newdocs.py [hunk 1]
```diff
@@ -3003,12 +3003,13 @@
...
 
-    Fill the array with a scalar value.
+    Fill the array with a value.
 
     Parameters
     ----------
     value : scalar
-        All elements of `a` will be assigned this value.
+        All elements of `a` will be assigned this value. If `a` is not a
+        (Python) object, `value` must be a scalar value.
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py]: How about "`value` must be a scalar unless `a` is an object array?

**Comment by eric-wieser** [comment]
I'm actually not sure TypeError _is_ appropriate here. I can pass in `np.array(0)` and that has the same type as `np.array([0])` which fails. So I'm inclined to close this. Thoughts, @seberg?

What might be cool is to add a new `class ShapeError(ValueError)` exception to numpy, which we can use in exactly this type of situation.

**Comment by seberg** [comment]
I would agree that disallowing 0D arrays completely would make sense (and they are pretty rare). That would require a slightly annoying deprecation though. As long as we let them through, I cannot say I care too much, but it probably is not be worth the trouble, than to be honest.

If ShapeError also encompasses BroadcastError's, it could make sense, I guess.

**Comment by mattip** [comment]
Closing. If someone feels strongly about it they can reopen with an updated PR

**Pull request closed without merging**